File: jena-patch/src/main/scala/eu/ostrzyciel/jelly/convert/jena/patch/JenaPatchConverterFactory.scala
package eu.ostrzyciel.jelly.convert.jena.patch

import eu.ostrzyciel.jelly.convert.jena.JenaConverterFactory
import eu.ostrzyciel.jelly.core.patch.PatchConverterFactory
import org.apache.jena.datatypes.RDFDatatype
import org.apache.jena.graph.Node

import scala.annotation.experimental

/**
 * Factory for Jena-based Jelly-Patch encoders and decoders.
 */
@experimental
object JenaPatchConverterFactory extends PatchConverterFactory[Node, RDFDatatype](
  JenaConverterFactory
)

File: jena-patch/src/main/java/eu/neverblink/jelly/convert/jena/patch/JenaPatchConverterFactory.java
package eu.neverblink.jelly.convert.jena.patch;

import eu.neverblink.jelly.convert.jena.JenaConverterFactory;
import eu.neverblink.jelly.convert.jena.JenaDecoderConverter;
import eu.neverblink.jelly.convert.jena.JenaEncoderConverter;
import eu.neverblink.jelly.core.ExperimentalApi;
import eu.neverblink.jelly.core.JellyConverterFactory;
import eu.neverblink.jelly.core.patch.JellyPatchConverterFactory;
import org.apache.jena.datatypes.RDFDatatype;
import org.apache.jena.graph.Node;

/**
 * Factory for Jena-based Jelly-Patch encoders and decoders.
 * <p>
 * This class is a singleton and should be accessed via the {@link #getInstance()} method.
 */
@ExperimentalApi
public final class JenaPatchConverterFactory
    extends JellyPatchConverterFactory<Node, RDFDatatype, JenaEncoderConverter, JenaDecoderConverter> {

    private static final JenaPatchConverterFactory INSTANCE = new JenaPatchConverterFactory(
        JenaConverterFactory.getInstance()
    );

    private JenaPatchConverterFactory(
        JellyConverterFactory<Node, RDFDatatype, JenaEncoderConverter, JenaDecoderConverter> converterFactory
    ) {
        super(converterFactory);
    }

    /**
     * Returns the singleton instance of the {@link JenaPatchConverterFactory}.
     *
     * @return the singleton instance of the {@link JenaPatchConverterFactory}
     */
    public static JenaPatchConverterFactory getInstance() {
        return INSTANCE;
    }
}
